### ISSUE TITLE
Support license-files [PEP-639]

### ISSUE THREAD

**Issue by sjlongland** [diazona/setuptools-pyproject-migration#77]
So, in issue #32 I was thrown off-course by the `setuptools` documentation, which documented both `license` _and_ `license_files` fields.

It was not clear whether the `license` field was sometimes read from a file instead of being verbatim text, or how the `license_files` influenced this field.

Experimental work suggested the two fields do not interact; a license file described in `license_files` does not yield text appearing in the `license` field.  Furthermore, `setuptools` does not support the `file:` prefix to denote text as coming from an external file as it does with other fields.

Having seen the library mentioned in issue #74 and reading the official Python package specification linked within (https://packaging.python.org/en/latest/specifications/core-metadata/#license), it is apparent this is a `setuptools` extension, undergoing discussion for inclusion under PEP-639 (https://peps.python.org/pep-0639/).

As such, we're probably better off waiting until a decision is made on PEP-639 before implementing it.

**Comment by diazona** [comment]
Yep, definitely makes sense to me to stick with the PEPs that have officially been accepted.

I suppose we could support 639 as an option, but that's not worth doing for our initial release IMO.

**Comment by diazona** [comment]
Looks like it's happening: https://floss.social/@hroncok/113044747383279536

**Comment by sjlongland** [comment]
Damn, it's almost been two years since I filed this.  Time flies.

Oookay, I had a quick look at the specification, I might get some time today to look at this.

**Comment by diazona** [comment]
Sounds good. Not that it needs to be done today of course... I was thinking I'd like to make the next release by the end of next month (so we don't go more than a year between releases) so any time by then works equally well as far as I'm concerned.

**Comment by sjlongland** [comment]
Had a closer look this afternoon… it appears `license-files` and `license` in the TOML file are related but parallel concepts:

https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license-and-license-files

`license` is supposed to meet the SPDX specification from here-on in… and we're supposed to [validate it](https://packaging.python.org/en/latest/specifications/pyproject-toml/#license):

> Tools SHOULD validate and perform case normalization of the expression.

`license-files` is a list of file globs, each one is supposed to [match at least one file](https://packaging.python.org/en/latest/specifications/pyproject-toml/#license-files).

This fits with the "not interacting" observation.  License files can include things like a list of authors which is not a software license.

My first thought is whether there's a library to validate the SPDX expression, since it'll likely change enough as licenses are added, we really don't want to be maintaining that.  So there's the next step: see what SPDX validation modules are out there.  `pypi.org` here I come.

**Comment by sjlongland** [comment]
* https://pypi.org/project/spdx-matcher/ -- has a nice feature in that it can extract a probable license from free-form text and suggest what might be the valid SPDX expression, however it requires Python 3.9, so would have to be an _optional_ dependency if we use it at all.
* https://pypi.org/project/spdx-lookup/ -- even works on Python 2.7 and early 3.x, so should be relatively safe, will implement the basic checking functionality.

The latter seems to support matching from arbitrary text too:

```python
# Find closest match for provided license content
with open('some-license.txt') as f:
    match = lookup.match(f.read()) # -> returns LicenseMatch or None
```

so maybe that's the one to use.

**Comment by sjlongland** [comment]
Does what it says on the tin…

```python3
(env) stuartl@rikishi ~/projects/setuptools-pyproject-migration/env $ python3
Python 3.11.13 (main, Jul  6 2025, 11:12:12) [GCC 14.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import spdx_lookup as lookup
>>> lookup.by_id('gpl-3.0')
{'name': 'GNU General Public License v3.0 only', 'id': 'GPL-3.0', 'sources': ['http://www.gnu.org/licenses/gpl-3.0-standalone.html', 'http://www.opensource.org/licenses/GPL-3.0'], 'notes': 'This license was released: 29 June 2007\nThis refers to when this GPL 3.0 only is being used (as opposed to "or later).', 'osi_approved': True, 'header': 'Copyright (C) <<var;name=copyright;original=<year>  <name of author>;match=.+>>[year]\n \nThis program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version.\n\nThis program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\nYou should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.', 'template': 'GPL-3.0.txt'}
>>> gpl3=lookup.by_id('gpl-3.0')
>>> gpl3.id
'GPL-3.0'
>>> lookup.by_id('made-up-and-invalid-spdx-expression')
# `None` returned
>>> gpl3.notes
'This license was released: 29 June 2007\nThis refers to when this GPL 3.0 only is being used (as opposed to "or later).'
>>> print(gpl3)
GNU General Public License v3.0 only
>>> print(gpl3.notes)
This license was released: 29 June 2007
This refers to when this GPL 3.0 only is being used (as opposed to "or later).
>>> print(gpl3.sources)
['http://www.gnu.org/licenses/gpl-3.0-standalone.html', 'http://www.opensource.org/licenses/GPL-3.0']
>>> print(gpl3.header)
Copyright (C) <<var;name=copyright;original=<year>  <name of author>;match=.+>>[year]
 
This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
>>> lookup.match(open("../work/LICENSE","r").read())
LicenseMatch(confidence=97.84946236559139, license={'name': 'MIT License', 'id': 'MIT', 'sources': ['http://www.opensource.org/licenses/MIT'], 'notes': None, 'osi_approved': True, 'header': None, 'template': 'MIT.txt'}, filename=None)
```

**Comment by sjlongland** [comment]
One thing it does _not_ handle, is composite licenses:

https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/#d4-composite-license-expressions

So some handling is needed.

**Comment by diazona** [comment]
Nice, thanks for sharing the research!

Off the top of my head, I think using an actively maintained project and trying to get the bugs in it fixed (or otherwise, working around them) is a better way to go. Maintainability counts for a lot. I'm totally fine with an optional dependency if that's what it takes. (Although dropping support for EOL versions of Python is always on the table if it makes sense, but using an optional dependency is an easy way to proceed that doesn't require fully dropping support for older Python versions, so I'd go with that.)

There is also the option of just writing our own code to do this. The [underlying data set](https://github.com/spdx/license-list-data) is available, and I bet some simple parsing logic could handle the vast majority of cases. That'd also have the advantage that we don't tie ourselves to Python 3.9+.

**Comment by sjlongland** [comment]
Yeah, I don't like having to re-invent the wheel but in this case, all the newer wheels are designed to fit newer axles than the ones we're trying to support.  Being able to match the license by the text was a _nice_ feature, but not essential, that'd be hard to re-create.

https://docs.python.org/3/library/importlib.resources.html documents a way to access package resources that are installed with the module, and this seems like a simple way out: drop the relevant `.json` files from that project into the appropriate spot in the source tree and have the code parse them on-the-fly.  `pyproject.toml` though specifies no common method for documenting what those data files are, apparently it's a [`setuptools` extension](https://setuptools.pypa.io/en/latest/userguide/datafiles.html) which we are using, so maybe that's the avenue to take if we do this.

**Comment by diazona** [comment]
Yeah, I think in general specifying which files get included in the sdist and wheel is a backend-specific feature. Different backends (flit, hatchling, pdm, etc.) have different ways of doing this - some might automatically include any data files that are there, others might never include data files (flit in particular is pretty bare-bones IIRC), others probably have configuration file entries which could live in the `tool.*` tables in `pyproject.toml`, and so on. Of course none of that makes a difference to how we'd actually load the file using `importlib.resources` at runtime. (Funnily enough, at work today I was actually implementing a resource loader)

That being said, I don't even think we need to do that, because

- [`spdx-license-list`](https://pypi.org/project/spdx-license-list/) exists, so we could just depend on that; it's probably a more reliably maintained dependency than either of the two parser packages you mentioned. But we should note that this would mean the version of the license list data we have available would depend on which Python version is being run; 3.7 and 3.8 wouldn't have access to more recent releases.
- I was also thinking it might make sense to download the list when needed, in order to always have the latest license data available. The disadvantage is that it doesn't work in an offline environment, although I'm not sure if that matters to anyone. If we want to support it, we could just include code for both, i.e. have an `offline` extra that depends on `spdx-license-list` (or one of those other parser packages), and then at runtime use the local data set if the dependency is installed, otherwise try to download it.

**Comment by sjlongland** [comment]
> Due to strong insistence from the license stewards for GNU licenses, we created separate license ID’s for the “only” and “or-later”.  These are not part of the expression syntax and therefore not processed by any of the machine readable SPDX license expression parsers – one would have to read the license notes to understand the semantics.  In other words, the “only” and “or-later” is a convention used by GNU that we carried forward in the license ID’s – not something intended to be standardized in the SPDX license syntax.

https://lists.spdx.org/g/spdx/message/1781

Thank-you FSF for making things simpler by introducing _additional_ distinct license identifiers for `-or-later` and `-only` variants of the same license.

So that settles it, `+` is the official "or later" method and can apply to any license (even unversioned ones, the meaning of "later" is dependent on the license) and the `-or-later` suffix should be considered a totally separate license for our purposes.

Of course the real WTF is that this is a legal solution to a social problem.

**Comment by diazona** [comment]
Ah gotcha, makes sense, thanks for explaining!

**Comment by sjlongland** [comment]
Been so many interruptions this year, barely got around to this.  But, no matter… I think I'm close.

Had a look tonight, and the logic is mostly there now.  `license_files` handling though had a few unexpected twists.  In the original `setuptools` docs, it's listed as a `list-comma`.  What that means in practice, is as a work-around to INI-file syntax limitations, it is given as a comma-separated string.

The code that does that is here:

https://github.com/pypa/setuptools/blob/d198e86f57231e83de87975c5c82bc40c196da79/setuptools/config/setupcfg.py#L307-L324

You'll note, it will also, parse a newline-separated list.  So we really need to handle both possibilities.

Helpfully, `setuptools` does *NOT* provide the result of this function: `self.distribution.command_options["metadata"]["license_files"]` actually is a tuple, the first element of which is the name of the `setuptools` config file that provided the value, and the _second_ is the raw string.

So we have to take the raw string, and parse that ourselves.  *Then* validate its contents.  So my simple "validate" became a "parse and validate", with the return value being the list of elements that were actually parsed out.

…and now for something completely different… indeed!

**Comment by sjlongland** [comment]
The other change now, I had a small change-of-heart regarding `spdx-lookup`.  It had one redeeming feature that was useful: the ability to match a license text to a license.

The legacy `setup.cfg` format does allow a full-license text to be embedded, and so in order to support this, `spdx-lookup` is used to find the most probably SPDX-compatible license that matches.  The heuristic is triggered if the license string is multi-line.

If a match is found, that is used with a warning emitted to the user suggesting they need to check the validity of this match, or copy the license out to its own file and specify it in `license_files`.

If no match is found, the script bails with a `ValueError`, suggesting they write the license out to its own file and reference it in `license_files`.

**Comment by diazona** [comment]
Awesome, thanks for the progress!

I'm fine dropping Python 3.7 support in order to get this in. Or you could try importing `TypedDict` from `typing_extensions`, if you want to make that effort.
